File: Behavioral/Lab14/C#/Entities/RealEstate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection.Emit;
using System.Text;
using System.Threading.Tasks;
using C_.Interfaces;

namespace C_.Entities
{
    public abstract class RealEstate : AgencyObject, ICloneable, IRentable, IBuyable, ISellable
    {
        public string Address { get; set; }
        public string Description { get; set; }
        public double Area { get; set; }
        public double PriceInUSD { get; set; }
        public string Status 
        { 
            get => _status != null ? _status : "Unknown";
            set
            {
                if (value == "New" || value == "Old" || value == "Middle aged")
                    _status = value;
                else
                    _status = "Unknown";
            } 
        }
        public RealtorAbstract Realtor{ get; set; }
        
        //Bools
        public bool IsRentable { get; set; } = false;
        public bool IsSellable { get; set; } = false;
        public bool IsBuyable { get; set; } = true;
        public bool Furnished { get; set; } = false;
        public bool Parking { get; set; } = false;
        public bool Balcony { get; set; } = false;
        public bool Elevator { get; set; } = false;
        public bool AirConditioning { get; set; } = false;
        public bool IsUtilitiesIncluded { get; set; } = false;
        
        private string _status;

        public RealEstate(string address, string description, double area, double price, string status):base() { 
            Address = address;
            Description = description;
            Area = area;
            PriceInUSD = price;
            Status = status;
            // Realtor = new Realtor("John Doe", "8 800 555 35 35", "example@gmail.com", 2);
        }

        public override string ToString(){
            return $"Id: {Id}, Address: {Address},\n Description: {Description},\nArea: {Area}, Price: {PriceInUSD}\nStatus: {Status}\n{Realtor}\nFurnished: {Furnished}, "+
            $"Parking: {Parking}, Balcony: {Balcony}, Elevator: {Elevator}, Air conditioning: {AirConditioning}, Utilities included: {IsUtilitiesIncluded}\n";
        }

        public abstract void Rent(Customer customer, DateTime startDate, DateTime endDate);
        public abstract void Sell(Customer customer);
        public abstract void SetEstateToBuy(Customer customer, double price);
        public abstract object Clone();
        public abstract RealEstate ShallowCopy();

        public static void PrintAllInfoAboutEstates(IEnumerable<RealEstate>  realEstates){
            foreach (RealEstate realEstate in realEstates){
                Console.WriteLine(realEstate.ToString());
            }
            Console.WriteLine("\n");
        }

        public virtual void Reset(){
            Address = "";
            Description = "";
            Area = 0;
            PriceInUSD = 0;
            Status = "";
            Realtor = null;
            IsRentable = false;
            IsBuyable = false;
            IsSellable = false;
            Furnished = false;
            Parking = false;
            Balcony = false;
            Elevator = false;
            AirConditioning = false;
            IsUtilitiesIncluded = false;
        }

    }
}

File: Behavioral/Lab23/Typescript/src/Entities/real_estate.ts
import { IBuyable } from '../Interfaces/buyable';
import { ISellable } from '../Interfaces/sellable';
import { ICloneable } from '../Interfaces/clonable';
import { IRentable } from '../Interfaces/rentable';
import { IPrintEstateInfo } from '../Interfaces/print_estate_info';
import { AgencyObject } from './agency_object';
import { Customer } from './customer';
import { Realtor } from './realtor';

export abstract class RealEstate extends AgencyObject implements ICloneable, IRentable, IBuyable, ISellable, IPrintEstateInfo {
    public address: string;
    public description: string;
    public area: number;
    public priceInUSD: number;
    public status: string;
    public realtor: Realtor;

    // Bools
    public isRentable: boolean = false;
    public isSellable: boolean = false;
    public isBuyable: boolean = true;
    public furnished: boolean = false;
    public parking: boolean = false;
    public balcony: boolean = false;
    public elevator: boolean = false;
    public airConditioning: boolean = false;
    public isUtilitiesIncluded: boolean = false;

    private _status: string;

    constructor(address: string, description: string, area: number, price: number, status: string) {
        super();
        this.address = address;
        this.description = description;
        this.area = area;
        this.priceInUSD = price;
        this.status = status;
    }

    public toString(): string {
        return `Id: ${this.id}, Address: ${this.address},\nDescription: ${this.description},\nArea: ${this.area}, Price: ${this.priceInUSD}\nStatus: ${this.status}\n${this.realtor}\nFurnished: ${this.furnished}, ` +
            `Parking: ${this.parking}, Balcony: ${this.balcony}, Elevator: ${this.elevator}, Air conditioning: ${this.airConditioning}, Utilities included: ${this.isUtilitiesIncluded}\n`;
    }
    public printEstateInfo(tab: string = ""): void {
        console.log(tab + this.toString().replace(/\n/g, "\n" + tab));
    }

    public abstract rent(customer: Customer, startDate: Date, endDate: Date): void;
    public abstract sell(customer: Customer): void;
    public abstract setEstateToBuy(customer: Customer, price: number): void;
    public abstract clone(): RealEstate;
    public abstract shallowCopy(): RealEstate;

    public static printAllInfoAboutEstates(realEstates: RealEstate[]): void {
        realEstates.forEach(realEstate => {
            console.log(realEstate.toString());
        });
        console.log("\n");
    }

    public reset(): void {
        this.address = "";
        this.description = "";
        this.area = 0;
        this.priceInUSD = 0;
        this.status = "";
        this.realtor = null;
        this.isRentable = false;
        this.isBuyable = false;
        this.isSellable = false;
        this.furnished = false;
        this.parking = false;
        this.balcony = false;
        this.elevator = false;
        this.airConditioning = false;
        this.isUtilitiesIncluded = false;
    }
}